### ISSUE TITLE
[Feature]: Add a 2-bit KV-cache quantisation backend from Together AI

### ISSUE THREAD

**Issue by zhangj1an** [vllm-project/vllm#46221]
## Purpose

Integrate a new INT2 KV-cache quantisation backend.

Paper:**OSCAR** ([arXiv:2605.17757](https://arxiv.org/abs/2605.17757)). 
Code: https://github.com/FutureMLS-Lab/OSCAR 
(500+ stars, proved to be working on Qwen3-32B, GLM-4.7-FP8.)
Website: https://oscar-quantize.github.io/

vLLM already has TurboQuant, a data-free Hadamard-rotation KV-cache quantizer. Oscar is similar, but has better accuracy. If we apply the most aggresive quantisation method in TurboQuant, (`turboquant_3bit_nc`), the accuracy will drop by 20%. But with Oscar, for Qwen3-32B and GLM-4.7-FP8 respectively, the mean gap to BF16 is −0.02, and +0.27 points while reducing KV memory by ~8× and increasing large-batch throughput by up to ~7×. 

The main idea is, before applying clipped INT2 scalar quantization, we first apply a per-layer covariance-aware orthogonal rotation to distribute the values evenly, so they can be later de-quantised with higher precision. 

To use this feature, just need to load the original model checkpoints, and download a small per-layer rotation file
(`[head_dim, head_dim]` orthogonal matrices, a few MB), published in the [OSCAR RotationZoo](https://huggingface.co/Zhongzhu/OSCAR-RotationZoo).

## CC List
Paper author: @zzz0906
vLLM maintainer: @mgoin, @robertgshaw2-redhat 

## Implementation Details
- We stored rotation matrix for K and V. 
- Then, we use the same fp16 `(scale, zero)` parameter for each vector. We then store these values by re-using TurboQuant's `TQFullAttentionSpec` in a Combined `[key_packed | value_packed]` slot (72 B/token/kv-head
  at `head_dim=128`). 
- We also introuced 3 additional triton kernels for this feature. Fused Triton store (quant + pack + scatter) and split-KV decode attention (INT2 unpack + online softmax + AV), plus a full-dequant kernel for the continuation-prefill path.

We will mainly change these files:
- `vllm/model_executor/layers/quantization/oscar/`: config,  rotation loading.
- `vllm/v1/attention/backends/oscar_attn.py`: prefill (BF16 attention + INT2 store), decode (rotate-Q → INT2 score/AV → `R_vᵀ` output).
- `vllm/v1/attention/ops/triton_oscar_{store,decode}.py`: fused Triton INT2 pack/store and split-KV decode-attention kernels.

Usage:

```bash
export VLLM_OSCAR_K_ROTATION_PATH=.../Qwen3-32B/.../k_rotation_qqt_r_h_pbr.pt
export VLLM_OSCAR_V_ROTATION_PATH=.../Qwen3-32B/.../v_rotation_sst_r_h_pbr.pt
export VLLM_OSCAR_K_CLIP_RATIO=0.96 VLLM_OSCAR_V_CLIP_RATIO=0.92
vllm serve Qwen/Qwen3-32B --kv-cache-dtype oscar_int2
```

## Reduce in KV-cache memory

As reported by paper, at `head_dim=128`, per token per KV head:

| Scheme | bytes | bits/elem | ratio |
|---|---|---|---|
| BF16 (K+V) | 512 | 16.0 | 1.0× |
| OSCAR INT2 slot | 72 | 2.25 | **7.11×** |
| OSCAR effective (4/64 layers BF16) | 99.5 | 3.11 | **5.15×** |

## Test plan

1. **Kernel correctness**: `tests/quantization/test_oscar.py`: INT2 store→dequant round-trip vs a PyTorch reference quantizer, and the fused INT2 decode kernel vs reference attention over the dequantized cache.
2. **End-to-end coherence**: Qwen3-32B, `--kv-cache-dtype oscar_int2` with the RotationZoo rotations.
3. **Accuracy**: GPQA-Diamond, BF16 vs OSCAR INT2, Qwen3-32B (thinking enabled).

## Not a duplicate

Checked before opening (please re-run; `gh` was unavailable in the dev sandbox):
```
gh pr list --repo vllm-project/vllm --state open --search "OSCAR"
gh pr list --repo vllm-project/vllm --state open --search "INT2 KV cache"
gh issue list --repo vllm-project/vllm --search "OSCAR 2-bit KV"
```
No existing OSCAR / INT2-KV-cache backend in `main`; the only adjacent work is TurboQuant (3-bit, data-free), which this complements at the 2-bit frontier.

### Alternatives

- Use `turboquant_3bit_nc`. It would incur ~50% larger KV than OSCAR INT2 and performs worth on reasoning at the 2-bit frontier, as stated by the paper.
- Per-tensor INT2 without calibrated rotation. The model will just collapse (verified on my local test. identity-rotation INT2 on Qwen3 produces incoherent text).

### Additional context

High accuracy, low-bit quantisation method is becoming increasing important in production use, such as autoround in `llm-compressor`.  So we believe this algo will be a valuable add. 

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by zhangj1an** [comment]
I have a PR ready and will create it tonight. I will paste benchmark results for lm eval and performance using Qwen3-32B as an example.  : )

**Comment by lloveapple** [comment]
> I have a PR ready and will create it tonight. I will paste benchmark results for lm eval and performance using Qwen3-32B as an example. : )

Nice! How is the accuracy of this algorithm?

**Comment by zzz0906** [comment]
Thanks for the issues & PR & integration. I will go through it too! I have tested multiple models across 6 benchmarks, including math, reasoning, agentic tasks, and ruler tasks up to 128K. The performance is quite good. Slight drop for qwen3-4B thinking and almost no drop for bigger models like Gemma-12B, qwen3-8,32B, GLM 4.7-FP8. For more accuracy results, please refer to https://github.com/FutureMLS-Lab/OSCAR. Thanks!

**Comment by zhangj1an** [comment]
Hey @lloveapple, thanks for your interest! The PR is ready in #46774. For accuracy, we will compare with existing vllm quantisation options (original fp16, nvfp4 and turboquant) in 1-2 days. If you have a particular dataset / model combi you want to see, please comment below, we will go and test it. Thank you!

**Comment by jagmarques** [comment]
Another calibration-free option in the 2-bit KV space: we built an E8 lattice VQ method (NexusQuant) that is calibration-free with Hadamard rotation. On Llama-3.1-8B-Instruct 4K (n=30 NIAH) our K2V2 at 2.0 bpe scores 30/30 while TurboQuant 2-bit collapses to 0/30 (McNemar b=29, c=0).

Validated on 9 architectures. pip install nexusquant-kv. Draft vLLM PR: #47742. Live demo: huggingface.co/spaces/jmarquex/nexusquant-demo
